pkg/v3/observer.go

```diff
@@ -27,54 +27,67 @@ type Runner interface {
 	CheckUpkeeps(context.Context, ...ocr2keepers.UpkeepPayload) ([]ocr2keepers.CheckResult, error)
 }
 
-type Observer struct {
-	Preprocessors []PreProcessor
-	Postprocessor PostProcessor
-	Runner        Runner
+type Observer[T any, R any] struct {
+	Preprocessors []PreProcessor[T]
+	Postprocessor PostProcessor[R]
+	processFunc   func(context.Context, ...T) ([]R, error)
 
 	// internal configurations
 	processTimeLimit time.Duration
 }
 
-// NewObserver creates a new Observer with the given pre-processors, post-processor, and runner
-func NewObserver(
-	preprocessors []PreProcessor,
-	postprocessor PostProcessor,
+// NewRunnableObserver creates a new Observer with the given pre-processors, post-processor, and runner
+func NewRunnableObserver(
+	preprocessors []PreProcessor[ocr2keepers.UpkeepPayload],
+	postprocessor PostProcessor[ocr2keepers.CheckResult],
 	runner Runner,
 	processLimit time.Duration,
-) *Observer {
-	return &Observer{
+) *Observer[ocr2keepers.UpkeepPayload, ocr2keepers.CheckResult] {
+	return &Observer[ocr2keepers.UpkeepPayload, ocr2keepers.CheckResult]{
 		Preprocessors:    preprocessors,
 		Postprocessor:    postprocessor,
-		Runner:           runner,
+		processFunc:      runner.CheckUpkeeps,
+		processTimeLimit: processLimit,
+	}
+}
+
+// NewGenericObserver creates a new Observer with the given pre-processors, post-processor, and runner
+func NewGenericObserver[T any, R any](
```

Comment:
The specific type of `RunnableObserver` is to indicate that it only accepts a `Runner` as the process function. I kinda struggled with the name on this one, but the idea is to enforce type safety and communicate the specific use case since that use case is very common.